File: language/runtime/apple/core/Sources/RuntimeHostAppleCore/Module/Intent/RuntimeHostIntent.swift
import Foundation

/// The Apple host intent payload delivered into one runtime session.
public enum RuntimeHostIntentPayload: Sendable, Hashable {
  /// One request to open one URL.
  case openURL(url: String)
  /// One request to open one file path.
  case openFile(path: String, contentType: String?)
  /// One request to deliver shared text.
  case shareText(text: String, contentType: String?)
  /// One request to deliver shared file paths.
  case shareFiles(paths: [String], contentType: String?)
  /// One custom action payload delivered by the Apple host.
  case customAction(
    action: String,
    url: String?,
    paths: [String],
    text: String?,
    contentType: String?
  )
}

/// One Apple intent ingress event delivered into one runtime session.
public struct RuntimeHostIntentEvent: Sendable, Hashable {
  /// The source bundle or process identifier when available.
  public let source: String?
  /// The intent payload delivered by the Apple host.
  public let payload: RuntimeHostIntentPayload

  /// Create one intent event.
  public init(
    source: String? = nil,
    payload: RuntimeHostIntentPayload
  ) {
    self.source = source
    self.payload = payload
  }
}

File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/module/intent/RuntimeHostIntent.kt
package dev.destack.runtime.android.module.intent

/**
 * The Android host intent payload delivered into one runtime session.
 */
public sealed interface RuntimeHostIntentPayload {
    /**
     * One request to open one URL.
     */
    public data class OpenUrl(
        /**
         * The URL delivered by the Android host.
         */
        val url: String,
    ) : RuntimeHostIntentPayload

    /**
     * One request to open one file path.
     */
    public data class OpenFile(
        /**
         * The path delivered by the Android host.
         */
        val path: String,

        /**
         * The normalized content type when available.
         */
        val contentType: String? = null,
    ) : RuntimeHostIntentPayload

    /**
     * One request to deliver shared text.
     */
    public data class ShareText(
        /**
         * The shared text payload.
         */
        val text: String,

        /**
         * The normalized content type when available.
         */
        val contentType: String? = null,
    ) : RuntimeHostIntentPayload

    /**
     * One request to deliver shared file paths.
     */
    public data class ShareFiles(
        /**
         * The shared file-path payloads.
         */
        val paths: List<String>,

        /**
         * The normalized content type when available.
         */
        val contentType: String? = null,
    ) : RuntimeHostIntentPayload

    /**
     * One custom action payload delivered by the Android host.
     */
    public data class CustomAction(
        /**
         * The custom action identifier.
         */
        val action: String,

        /**
         * The optional URL payload.
         */
        val url: String? = null,

        /**
         * The optional file-path payloads.
         */
        val paths: List<String> = emptyList(),

        /**
         * The optional shared text payload.
         */
        val text: String? = null,

        /**
         * The normalized content type when available.
         */
        val contentType: String? = null,
    ) : RuntimeHostIntentPayload
}

/**
 * One Android intent ingress event delivered into one runtime session.
 */
public data class RuntimeHostIntentEvent(
    /**
     * The source package or process identifier when available.
     */
    val source: String? = null,

    /**
     * The intent payload delivered by the Android host.
     */
    val payload: RuntimeHostIntentPayload,
)
